cgroup/fanotify.go

```diff
@@ -0,0 +1,205 @@
+				return fmt.Errorf("error reading extra stuff: %v", err)
+			}
+		}
+
+		if err := m.handleFanotify(&metadata, buf[:size]); err != nil {
+			return fmt.Errorf("error handling fanotify event: %v", err)
+		}
+	}
+}
+
+func (m *fanotifyMonitor) handleFanotify(metadata *unix.FanotifyEventMetadata, buf []byte) error {
+	reader := bufio.NewReader(bytes.NewReader(buf))
+
+	header := fanotifyEventInfoFid{}
+	if err := binary.Read(reader, m.byteOrder, &header); err != nil {
+		return fmt.Errorf("error reading fanotify header: %v", err)
+	}
+
+	if header.Header.Type != unix.FAN_EVENT_INFO_TYPE_DFID_NAME {
+		return fmt.Errorf("unexpected fanotify header type 0x%x (expected 0x%x / FAN_EVENT_INFO_TYPE_DFID_NAME)", header.Header.Type, unix.FAN_EVENT_INFO_TYPE_DFID_NAME)
+	}
+
+	handle, err := m.readFanotifyFileHandle(reader)
+	if err != nil {
+		return fmt.Errorf("error reading file_handle: %v", err)
+	}
+
+	dir, err := m.resolveDir(handle)
+	if err != nil {
+		return fmt.Errorf("error reading fanotify event path: %v", err)
+	}
+
+	name, err := reader.ReadString('\x00')
+	if err != nil {
+		return fmt.Errorf("error reading name: %v", err)
+	}
+
+	// Truncate \x00 at the end
+	name = name[:len(name)-1]
+
+	path := fmt.Sprintf("%s/%s", dir, name)
+
+	inode, err := inode(path)
+	if err != nil {
+		return fmt.Errorf("error resolving inode for %q: %v", path, err)
+	}
+
+	m.mapping[inode] = path
+
+	return nil
+}
+
+func (m *fanotifyMonitor) resolveDir(handle unix.FileHandle) (string, error) {
+	fd, err := unix.OpenByHandleAt(int(m.mount.Fd()), handle, 0)
+	if err != nil {
+		return "", fmt.Errorf("error opening event fd: %v", err)
+	}
+
+	defer syscall.Close(fd)
+
+	dir, err := filepath.EvalSymlinks(fmt.Sprintf("/proc/self/fd/%d", fd))
+	if err != nil {
+		return "", fmt.Errorf("error resolving event fd symlink: %v", err)
+	}
+
+	return dir, nil
+}
+
+// Reading struct file_handle piece by piece:
+// https://elixir.bootlin.com/linux/v6.5-rc1/source/include/linux/fs.h#L1002
+func (m *fanotifyMonitor) readFanotifyFileHandle(reader io.Reader) (unix.FileHandle, error) {
+	handleBytes := uint32(0)
+	if err := binary.Read(reader, binary.LittleEndian, &handleBytes); err != nil {
+		return unix.FileHandle{}, fmt.Errorf("error reading file_handle->handle_bytes: %v", err)
+	}
+
+	handleType := int32(0)
+	if err := binary.Read(reader, binary.LittleEndian, &handleType); err != nil {
+		return unix.FileHandle{}, fmt.Errorf("error reading file_handle->handle_type: %v", err)
+	}
+
+	handle := make([]byte, handleBytes)
+	if _, err := reader.Read(handle); err != nil {
+		return unix.FileHandle{}, fmt.Errorf("error reading file_handle->handle (%d bytes): %v", handleBytes, err)
+	}
+
+	return unix.NewFileHandle(handleType, handle), nil
+}
+
+func (m *fanotifyMonitor) Resolve(id int) string {
+	return m.mapping[id]
+}
+
+// The following kernel patch is required to take advantage of this:
+// * https://lore.kernel.org/linux-fsdevel/20230731184731.64568-1-ivan@cloudflare.com/
```

Comment:
Linux v6.6-rc1 is out with the following commit:

* https://github.com/torvalds/linux/commit/0ce7c12e88cfa3e4cb0800e8172e5befea5c199a

I need to rebase and update the link, and then it should be good to go.